pkg/controllers/workgenerator/reconciler.go

```diff
@@ -0,0 +1,299 @@
+		return nil, controller.NewAPIServerError(true, err)
+	}
+	resourceSnapshots[0] = &resourceSnapshot
+
+	// check if there are more snapshot in the same index group
+	countAnnotation := resourceSnapshot.Annotations[fleetv1beta1.NumberOfResourceSnapshotsAnnotation]
+	snapshotCount, err := strconv.Atoi(countAnnotation)
+	if err != nil {
+		return nil, controller.NewUnexpectedBehaviorError(err)
+	}
+	if snapshotCount > 1 {
+		index, exist := resourceSnapshot.Labels[fleetv1beta1.ResourceIndexLabel]
+		if !exist {
+			unexpectedErr := fmt.Errorf("resource snapshot %s has no index label", resourceSnapshot.Name)
+			return nil, controller.NewUnexpectedBehaviorError(unexpectedErr)
+		}
+		resourceIndexLabelMatcher := client.MatchingLabels{
+			fleetv1beta1.ResourceIndexLabel: index,
+		}
+		resourceSnapshotList := &fleetv1beta1.ClusterResourceSnapshotList{}
+		if err := r.Client.List(ctx, resourceSnapshotList, resourceIndexLabelMatcher); err != nil {
+			klog.ErrorS(err, "failed to list all the resource snapshot associated with the resourceBinding", "resourceBinding", klog.KObj(resourceBinding))
+			return nil, controller.NewAPIServerError(true, err)
+		}
+		for i := 0; i < snapshotCount; i++ {
+			resourceSnapshots = append(resourceSnapshots, &resourceSnapshotList.Items[i])
+		}
+	}
+	return resourceSnapshots, nil
+}
+
+// upsertWork creates or updates the work for the corresponding resource snapshot
+func (r *Reconciler) upsertWork(ctx context.Context, work *workv1alpha1.Work, resourceSnapshot *fleetv1beta1.ClusterResourceSnapshot, resourceBinding *fleetv1beta1.ClusterResourceBinding) error {
+	needCreate := false
+	if work == nil {
+		needCreate = true
+		work = &workv1alpha1.Work{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      resourceSnapshot.Name, // work name same as the corresponding resource snapshot name
+				Namespace: fmt.Sprintf(utils.NamespaceNameFormat, resourceBinding.Spec.TargetCluster),
+				Labels: map[string]string{
+					fleetv1beta1.ParentBindingLabel: resourceBinding.Name,
+					fleetv1beta1.CRPTrackingLabel:   resourceBinding.Labels[fleetv1beta1.CRPTrackingLabel],
+				},
+				OwnerReferences: []metav1.OwnerReference{
+					{
+						APIVersion:         fleetv1beta1.GroupVersion.String(),
+						Kind:               resourceBinding.Kind,
+						Name:               resourceBinding.Name,
+						UID:                resourceBinding.UID,
+						BlockOwnerDeletion: pointer.Bool(true), // make sure that the k8s will call work delete when the binding is deleted
+					},
+				},
+			},
+		}
+	}
+	workObj := klog.KObj(work)
+	resourceBindingObj := klog.KObj(resourceBinding)
+	// reset the manifest list regardless
+	work.Spec.Workload.Manifests = make([]workv1alpha1.Manifest, 0)
+	for _, selectedResource := range resourceSnapshot.Spec.SelectedResources {
+		work.Spec.Workload.Manifests = append(work.Spec.Workload.Manifests, workv1alpha1.Manifest{
+			RawExtension: selectedResource.RawExtension,
+		})
+	}
+
+	// upsert the work
```

Comment:
Hi Ryan! controller runtime package has this CreateOrUpdate method which may be helpful: https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/controller/controllerutil#CreateOrUpdate.